src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs

```diff
@@ -431,28 +432,116 @@ private static bool OverloadResolutionResultIsValid<TMember>(ArrayBuilder<Member
         }
 
 #nullable enable
-        private void RemoveCallingConventionMismatches<TMember>(ArrayBuilder<MemberResolutionResult<TMember>> results, Cci.CallingConvention expectedConvention) where TMember : Symbol
+        private void RemoveCallingConventionMismatches<TMember>(ArrayBuilder<MemberResolutionResult<TMember>> results, in CallingConventionInfo expectedConvention) where TMember : Symbol
         {
             if (typeof(TMember) != typeof(MethodSymbol))
             {
                 return;
             }
 
+            Debug.Assert(expectedConvention.CallingConventionTypes is not null);
+
             for (int i = 0; i < results.Count; i++)
             {
                 var result = results[i];
                 var member = (MethodSymbol)(Symbol)result.Member;
                 if (result.Result.IsValid)
                 {
-                    if (!member.CallingConvention.IsCallingConvention(expectedConvention)
-                        || member.CallingConvention.HasUnknownCallingConventionAttributeBits())
+                    if (member.CallingConvention.HasUnknownCallingConventionAttributeBits())
+                    {
+                        results[i] = makeWrongCallingConvention(result);
+                        continue;
+                    }
+
+                    var unmanagedCallersOnlyData = member.UnmanagedCallersOnlyAttributeData;
+                    if (ReferenceEquals(unmanagedCallersOnlyData, UnmanagedCallersOnlyAttributeData.AttributePresentDataNotBound)
+                        || ReferenceEquals(unmanagedCallersOnlyData, UnmanagedCallersOnlyAttributeData.Uninitialized))
+                    {
+                        // If we're at a location where the unmanaged data has actually not yet been bound, it cannot be
+                        // valid code anyway, as the only place that can occur is when we're actually doing an address of
+                        // in an attribute.
+                        continue;
+                    }
+
+                    if (unmanagedCallersOnlyData is null)
+                    {
+                        if (!member.CallingConvention.IsCallingConvention(expectedConvention.CallKind)
+                            || !expectedConvention.CallingConventionTypes.IsEmpty)
```

Comment:
>!expectedConvention.CallingConventionTypes.IsEmpty [](start = 31, length = 50)

It is not clear why presence of types necessary makes candidate unused. Is the assumption that ```member.CallingConvention``` is never unmanaged and presence of types implies that the target is unmanaged?

Reply:
For regular method symbols, we do not decode the extensible calling convention modopts when determining the calling convention. So, have one of 2 cases:
1. The expected convention is one of the well-known ones that maps directly to a CallKind, in which case the previous check is what we need
2. If the member's CallKind is Unmanaged and the expected is Unmanaged, we need to additionally check to make sure that the expected calling convention is not an extension convention. If we expected an extension convention that is not one of the ones that is converted to a CallKind, then those types will be in this set and there is an error. I'll rename `CallingConventionTypes` to `UnmanagedCallingConventionTypes` to hopefully make this a bit clearer.

---
In reply to: [488976731](https://github.com/dotnet/roslyn/pull/47651#discussion_r488976731) [](ancestors = 488976731)